File: day3/main.py
from pathlib import Path

lines = Path("./input").read_text().rstrip("\n").split("\n")
lines = [list(x) for x in lines]
lines = [[int(y) for y in x] for x in lines]

n = len(lines)

# Part 1
gamma = [str(int(sum([x[i] for x in lines]) / n >= 0.5)) for i in range(len(lines[0]))]
epsilon = [str(int(sum([x[i] for x in lines]) / n <= 0.5)) for i in range(len(lines[0]))]

gamma = int("".join(gamma), 2)
epsilon = int("".join(epsilon), 2)

print(gamma*epsilon)

# Part 2
O_lines = list(lines)
CO2_lines = list(lines)

for i in range(len(lines[0])):
    if len(O_lines) > 1:
        most_common = int(sum([x[i] for x in O_lines]) / len(O_lines) >= 0.5)
        O_lines = list(filter(lambda x: most_common == x[i], O_lines))
    if len(CO2_lines) > 1:
        least_common = int(sum([x[i] for x in CO2_lines]) / len(CO2_lines) < 0.5)
        CO2_lines = list(filter(lambda x: least_common == x[i], CO2_lines))
    if len(O_lines) == 1:
        O_rating = int("".join([str(x) for x in O_lines[0]]), 2)
    if len(CO2_lines) == 1:
        CO2_rating = int("".join([str(x) for x in CO2_lines[0]]), 2)

print(O_rating*CO2_rating)

File: day3/main.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func readFile(file string) ([]int, int) {
	content, err := os.ReadFile(file)
	if err != nil {
		log.Fatal(err)
	}
	lines := strings.Split(strings.Trim(string(content), "\n"), "\n")

	num_bits := len(lines[0])
	num_items := len(lines)
	nums := make([]int, num_items)

	for i, x := range lines {
		num, err := strconv.ParseInt(x, 2, 32)
		if err != nil {
			log.Fatal(err)
		}
		nums[i] = int(num)
	}
	return nums, num_bits
}

func calcMajorityBit(nums []int, place int) int {
	sum := 0
	for _, x := range nums {
		sum += x >> place & 1
	}
	if (float64(sum) / float64(len(nums))) >= 0.5 {
		return 1
	} else {
		return 0
	}

}

func filterCO2Nums(nums []int, place int) []int {
	majorityBit := calcMajorityBit(nums, place)
	var result []int
	for _, x := range nums {
		if majorityBit != ((x >> place) & 1) {
			result = append(result, x)
		}
	}
	return result
}

func filterOxygenNums(nums []int, place int) []int {
	majorityBit := calcMajorityBit(nums, place)
	var result []int
	for _, x := range nums {
		if majorityBit == ((x >> place) & 1) {
			result = append(result, x)
		}
	}
	return result
}

func part1(nums []int, num_bits int) int {
	gamma := 0
	epsilon := 0
	for i := 0; i < num_bits; i++ {
		place := num_bits - i - 1
		majorityBit := calcMajorityBit(nums, place)

		gamma |= majorityBit << place
		epsilon |= (majorityBit ^ 1) << place
	}
	return gamma * epsilon
}

func part2(nums []int, num_bits int) int {

	oxygenNums := make([]int, len(nums))
	copy(oxygenNums, nums)
	for i := num_bits - 1; i >= 0; i-- {
		oxygenNums = filterOxygenNums(oxygenNums, i)
		if len(oxygenNums) == 1 {
			break
		}
	}

	CO2Nums := make([]int, len(nums))
	copy(CO2Nums, nums)
	for i := num_bits - 1; i >= 0; i-- {
		CO2Nums = filterCO2Nums(CO2Nums, i)
		if len(CO2Nums) == 1 {
			break
		}
	}

	return oxygenNums[0] * CO2Nums[0]
}

func main() {
	nums, num_bits := readFile("input")
	fmt.Println(part1(nums, num_bits))
	fmt.Println(part2(nums, num_bits))
}
